presto-main/src/main/java/com/facebook/presto/sql/relational/RawExpressionUtils.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.relational;
+
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.RowExpressionVisitor;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.tree.Expression;
+
+import java.util.Objects;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * The following has a list of helper functions to transform between Expression and RowExpression.
+ * Ideally, users should never do type checking or down casting for future convenience of refactoring.
+ */
+public final class RawExpressionUtils
+{
+    private RawExpressionUtils() {}
+
+    /**
+     * Create a new RowExpression
+     */
+    public static RowExpression toRaw(Expression expression)
+    {
+        return new RawExpression(expression);
+    }
+
+    /**
+     * Degrade to Expression
+     */
+    public static Expression toExpression(RowExpression rowExpression)
+    {
+        checkArgument(isRaw(rowExpression));
+        return ((RawExpression) rowExpression).getExpression();
+    }
+
+    /**
+     * Check if the given {@param rowExpression} is a raw expression.
+     */
+    public static boolean isRaw(RowExpression rowExpression)
+    {
+        return (rowExpression instanceof RawExpression);
+    }
+
+    /**
+     * RawExpression is a RowExpression container holding an {@param Expression} object
+     * that cannot be translated directly while constructing PlanNode.
+     * Typical examples are cases with `WITH` clauses or alias that the type or arguments information
+     * are not directly available given the substrees are not formed yet.
+     * RawExpression should not exist after optimization or serialized over the wire.
+     * All RawExpression should be translated to other corresponding RowExpression objects ultimately.
+     */
+    private static final class RawExpression
```

Comment:
@highker RawExpression name is so similar to RowExpression that it is confusing. It is hard to distinguish when reading code and impossible when discussing it verbally. How would you feel about renaming it to something like UnresolvedRowExpression, PendingRowExpression or FutureRowExpression?

Reply:
`UnresolvedRowExpression` might work. @rongrong what do you think?

Reply:
It's technically not "unresolved" and we don't call the conversion between Expression and RowExpression "resolve". But it sort of makes sense. What about `OriginalExpression`...?

Reply:
Or `ASTExpression`?

Reply:
I'd prefer `OriginalExpression` (or `SourceExpression` or `SqlExpression`). It is easier to pronounce than `ASTExpression`.